torchmultimodal/utils/generate.py

```diff
@@ -0,0 +1,277 @@
+            Defaults to ``1``.
+
+    Code reference: https://gist.github.com/thomwolf/1a5a29f6962089e871b94cbd09daf317
+
+    Args:
+        logits (Tensor): Logits distribution shape ``(b, num_tokens)`` where ``b`` is batch size,
+            ``num_tokens`` is the number of tokens.
+
+    Returns:
+        Filtered logits tensor.
+
+    Raises:
+        ValueError: If 'top_k' or 'top_p' are outside of valid numerical ranges.
+    """
+
+    def __init__(
+        self,
+        top_k: Optional[int] = None,
+        top_p: Optional[float] = None,
+        min_tokens_to_keep: int = 1,
+        filter_value: float = -float("inf"),
+    ) -> None:
+        if top_k is not None and top_k < 0:
+            raise ValueError(f"'top_k' must be non-negative but got {top_k}.")
+        if top_p is not None and (top_p > 1.0 or top_p < 0.0):
+            raise ValueError(f"'top_p' must be within `[0.0, 1.0]` but got {top_p}.")
+
+        self.min_tokens_to_keep = min_tokens_to_keep
+        self.filter_value = filter_value
+        self.top_k = top_k
+        self.top_p = top_p
+
+    def __call__(self, logits: Tensor) -> Tensor:
+        if self.top_k is not None:
+            logits = self._filter_top_k(logits)
+        if self.top_p is not None:
+            logits = self._filter_top_p(logits)
+
+        return logits
+
+    def _filter_top_k(self, logits: Tensor) -> Tensor:
+        if self.top_k == 0:
+            return logits
+
+        top_k = min(
+            max(self.top_k, self.min_tokens_to_keep), logits.size(-1)
+        )  # Safety check
+        # Remove all tokens with a probability less than the last token of the top-k
+        indices_to_remove = logits < torch.topk(logits, top_k)[0][..., -1:]  # (b, 1)
+        logits[indices_to_remove] = self.filter_value
+
+        return logits
+
+    def _filter_top_p(self, logits: Tensor) -> Tensor:
+        if self.top_p == 1.0:
+            return logits
+
+        sorted_logits, sorted_indices = torch.sort(logits, descending=True)
+        cumulative_probs = torch.cumsum(F.softmax(sorted_logits, dim=-1), dim=-1)
+
+        # Remove tokens with cumulative probability above the threshold (token with 0 are kept)
+        sorted_indices_to_remove = cumulative_probs > self.top_p
+        if self.min_tokens_to_keep > 1:
+            # Keep at least min_tokens_to_keep (set to min_tokens_to_keep-1 because we add
+            # the first one below)
+            sorted_indices_to_remove[..., : self.min_tokens_to_keep] = 0
+        # Shift the indices to the right to keep also the first token above the threshold
+        sorted_indices_to_remove[..., 1:] = sorted_indices_to_remove[..., :-1].clone()
+        sorted_indices_to_remove[..., 0] = 0
+
+        # scatter sorted tensors to original indexing
+        indices_to_remove = sorted_indices_to_remove.scatter(
+            1, sorted_indices, sorted_indices_to_remove
+        )
+        logits[indices_to_remove] = self.filter_value
```

Comment:
This feels needlessly complicated, can we just apply top-k and top-p separately? Torch already has a builtin for top-k [here](https://pytorch.org/docs/stable/generated/torch.topk.html), and top-p should also just be a simple filter

Reply:
By "separately" do you mean to split this into two classes one for `top_k` one for `top_p`?

While that's possible but in regarding your second point:

We are calling pytorch's `top_k` underneath but you would still need 1) sanity checking to ensure minimum number of tokens still exist after the filtering 2) assign `max_neg_value` to the unwanted tokens (so they are excluded from sampling without changing the size of the logit matrix).

https://github.com/facebookresearch/multimodal/blob/6ca2b5e2147ed3aaae996208c8b8e028416c34c8/torchmultimodal/utils/generate.py#L245-L250

Reply:
>top-p should also just be a simple filter

I am not aware if there's an out-of-box top-p filter we can use here.

Reply:
Thanks for the additional information here. I am still envisioning a case where users can pass both top-k and top-p, but that each has its own function (rather than being combined into a single class). I think this is ultimately what HF does anyways: [ref](https://github.com/huggingface/transformers/blob/169b8cde471ec48e557f4c2d0c8e0d40abb13040/src/transformers/generation_logits_process.py#L212), but as you mentioned they do add a utility function to combine each strategy. Regarding your suggestion:

> We can create individual classes for different filtering strategies
We still need a topping layer to manage the combination of strategies based on user inputs

Why do we need a separate class for managing combinations? Seems it could just be handled directly in the generation class (which already has access to the `top_k` and `top_p` values). Then imo the behavior is clearer anyways, we can see directly from the main class the roles of `top_k` and `top_p`, whereas otherwise they are obscured by the `LogitsFilter` class.

Reply:
I agree the selection logic fits more naturally to `GenerateUtil` given the latter already has the knobs (i.e. ,`top_k`, `top_p`). So:
 
 - Move the selection logic into `GenerateUtil` exposing the "knobs"
 - Convert `_filter_top_k` and `_filter_top_p` as two separate classes
 - `top_p` we will keep the current implementation as there's no out-of-box tool from pytorch core

let me know if we are in line now.